packages/zoe/src/zoe.chainmail

```diff
@@ -42,60 +47,52 @@ interface ZoeService {
 
   /** 
-   * To redeem an invite, the user must provide a list of payments as well as
-   * their rules for the offer. 
+   * To redeem an invite, the user must provide their rules for the
+   * offer as well as payments to be escrowed by Zoe.
    * 
-   * The rules for the offer are in two parts: `payoutRules` are used 
-   * by Zoe to enforce offer safety, and `exitRule` is used to specify
+   * The rules for the offer are in three parts: `want` and `offer` are used 
+   * by Zoe to enforce offer safety, and `exit` is used to specify
    * the extent of payout liveness that Zoe can guarantee.
-   * `payoutRules` is a list of objects, each with a `kind` property 
-   * (such as 'offerAtMost') and an amount property. The objects in the 
-   * `payoutRules` must be in the same order
-   * as the issuers associated with a smart contract. That is, the
-   * amount in index 0 of `payoutRules` should be an amount for the issuer
-   * in index 0 of the issuers array. `payments` is an array of the
-   * actual payments to be escrowed, following the rules in the
-   * payoutRules. If the payoutRules kind is 'offerAtMost',
-   * then a payment matching the amount is expected. If the payoutRules
-   * kind is 'wantAtLeast' then the payment will be
-   * ignored and should be `undefined`.
+   * `want` and `offer` are objects with roleNames as keys and amounts
+   * as values. `payments` is an object with the rolesNames as keys,
+   * and the values are the actual payments to be escrowed. A payment
+   * is expected for every rule under `offer`.
    */
-  redeem (invite :Invite, offerRules :OfferRules, payments :List(Payment)) -> (SeatAndPayout);
+  redeem (invite :Invite, offerRules :OfferRules, payments :PaymentByRole)) -> (SeatAndPayout);
 }
 
 /**
  * This is returned by a call to `redeem` on Zoe.
  * A seat is an arbitrary object whose methods allow the user to take
- * certain actions in a contract. The payout is a promise that resolves to an array
- * of promises for payments. Note that while the payout promise
- * resolves when an offer is completed, the promises for each payment
- * resolves after the remote issuer successfully withdraws the payment.
+ * certain actions in a contract. The payout is a promise that resolves
+ * to an object which has roleNames as keys and promises for payments
+ * as values. Note that while the payout promise resolves when an offer
+ * is completed, the promise for each payment resolves after the remote
+ * issuer successfully withdraws the payment.
  */
 struct SeatAndPayout ( ) {
   seat :Object;
   payout :List(Payment);
 }
 
 struct OfferRules ( ) {
-  payoutRules :List(PayoutRule);
-  exitRule :ExitRule;
+  want :RoleAmountObject;
+  offer :RoleAmountObject;
+  exit :ExitRule;
 }
 
 /**
- * payoutRules are an array of PayoutRule. The possible
- * kinds are 'offerAtMost' and 'wantAtLeast'.
+ * The keys are roleNames, and the values are amounts. For example:
+ * { Asset: moola(5), Price: simoleans(9) }
```

Comment:
I don't think `moola(5)` is sufficiently clear in this context.

Reply:
Ok, I will replace with AmountMath.make